<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>mcRigor: a statistical tool to enhance the rigor of metacell partitioning • mcRigor</title>
<script src="deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="deps/headroom-0.11.0/headroom.min.js"></script><script src="deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="deps/search-1.0.0/fuse.min.js"></script><script src="deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="pkgdown.js"></script><meta property="og:title" content="mcRigor: a statistical tool to enhance the rigor of metacell partitioning">
<meta name="description" content="Detection of dubiuos metacells and optimization of metacell partitioning.">
<meta property="og:description" content="Detection of dubiuos metacells and optimization of metacell partitioning.">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="index.html">mcRigor</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="reference/index.html">Reference</a></li>
<li class="nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="articles/mcRigor-1-detect-dubmc.html">Functionality 1: detect dubious metacells for a given metacell partition</a></li>
    <li><a class="dropdown-item" href="articles/mcRigor-2-optimize.html">Functionality 2: optimize metacell partitioning</a></li>
    <li><a class="dropdown-item" href="articles/mcRigor-3-mc-method.html">Implementing metacell partitioning methods</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-home">
<div class="row">
  <main id="main" class="col-md-9"><div class="section level1">
<div class="page-header"><h1 id="mcrigor">mcRigor<a class="anchor" aria-label="anchor" href="#mcrigor"></a>
</h1></div>
<p>The R package <strong>mcRigor</strong> is a statistical method to enhance the rigor of metacell partitioning in single-cell data analysis. It can be used as an add-on for any existing metacell partitioning methods for obtaining more reliable metacells. mcRigor has two main functionalities: 1) detecting dubious metacells, which are composed of heterogeneous single cells, for a given metacell partition, and 2) optimizing the hyperparameter of a metacell partitioning method. The core of mcRigor is a feature-correlation-based statistic that measures the heterogeneity of a metacell, with its null distribution derived from a double permutation scheme. The following figure illustrates the schematics of mcRigor for dubious metacell detection (a) and hyperparameter optimization (b).</p>
<p><img src="reference/figures/mcRigor_illustration.png" width="750"></p>
<div class="section level2">
<h2 id="installation">Installation<a name="installation-"></a>
<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>To install the development version from GitHub, please run:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="st"><a href="https://devtools.r-lib.org/" class="external-link">"devtools"</a></span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"devtools"</span><span class="op">)</span></span>
<span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"JSB-UCLA/mcRigor"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="quick-start">Quick Start<a name="quick-start"></a>
<a class="anchor" aria-label="anchor" href="#quick-start"></a>
</h2>
<p>The following code demonstrates how to implement the two main functionalities of <strong>mcRigor</strong> using its corresponding functions: <code>mcRigor_DETECT</code> and <code>mcRigor_OPTIMIZE</code>. For both functions, the input data should include:</p>
<ul>
<li>A Seurat object containing the raw single cell sequencing data.</li>
<li>A dataframe specifying the metacell partitions to rectify or optimize.</li>
</ul>
<p>An important difference between the two functions lies in their requirements of the second input: <code>mcRigor_DETECT</code> requires only a single metacell partition as input since its purpose is to detect dubious metacells within the given partition, while <code>mcRigor_OPTIMIZE</code> requires a series of metacell partitions as input since its goal is to identify the optimal partition from the provided candidates. Note that the metacell partitions can be obtained using existing metacell partitioning methods (e.g., SEACells, MetaCell, MetaCell2, SuperCell) or ad-hoc approaches. An example input is as below:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'syn.rds'</span>, package <span class="op">=</span> <span class="st">'mcRigor'</span><span class="op">)</span></span>
<span><span class="va">obj_singlecell</span><span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">sc_dir</span><span class="op">)</span>  <span class="co"># A Seurat object containing the raw scRNA-seq data</span></span>
<span></span>
<span><span class="va">membership_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'seacells_cell_membership_rna_syn.csv'</span>, package <span class="op">=</span> <span class="st">'mcRigor'</span><span class="op">)</span></span>
<span><span class="va">cell_membership_all</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">membership_dir</span>, check.names <span class="op">=</span> <span class="cn">F</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span>  <span class="co"># A dataframe containing a series of candidate metacell partitions to optimize.</span></span>
<span><span class="va">cell_membership</span> <span class="op">&lt;-</span> <span class="va">cell_membership_all</span><span class="op">[</span><span class="st">'50'</span><span class="op">]</span>  <span class="co"># A dataframe containing one metacell partition to rectify.</span></span></code></pre></div>
<p>We implement the first functionality: <strong>detecting dubious metacells</strong>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">detect_res</span> <span class="op">=</span> <span class="fu"><a href="reference/mcRigor_DETECT.html">mcRigor_DETECT</a></span><span class="op">(</span>obj_singlecell <span class="op">=</span> <span class="va">obj_singlecell</span>, </span>
<span>                            cell_membership <span class="op">=</span> <span class="va">cell_membership</span>,</span>
<span>                            tgamma <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>                            assay_type <span class="op">=</span> <span class="st">'RNA'</span>,</span>
<span>                            feature_use <span class="op">=</span> <span class="fl">2000</span>,</span>
<span>                            aggregate_method <span class="op">=</span> <span class="st">'mean'</span>,</span>
<span>                            Nrep <span class="op">=</span> <span class="fl">1</span>,</span>
<span>                            test_cutoff <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>                            thre_smooth <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>                            thre_bw <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">6</span><span class="op">)</span></span>
<span><span class="va">detect_res</span><span class="op">$</span><span class="va">mc_res</span></span></code></pre></div>
<p>The arguments of <code><a href="reference/mcRigor_DETECT.html">mcRigor_DETECT()</a></code> include:</p>
<ul>
<li>
<code>obj_singlecell</code>: A Seurat object of the original single cell sequencing data.</li>
<li>
<code>cell_membership</code>: A dataframe, each column of which contains the metacell membership of single cells under a given <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>γ</mi><annotation encoding="application/x-tex">\gamma</annotation></semantics></math> (granularity level). The column names should be the corresponding <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>γ</mi><annotation encoding="application/x-tex">\gamma</annotation></semantics></math>’s (in the character type). The row names should be the single cell ids.</li>
<li>
<code>tgamma</code>: An integer ofthe target <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>γ</mi><annotation encoding="application/x-tex">\gamma</annotation></semantics></math> value — the <math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>γ</mi><annotation encoding="application/x-tex">\gamma</annotation></semantics></math> (in the character type) corresponding to the metacell partition where dubious metacells detection will be performed. tgamma has to be in the column names of cell_membership. If tgamma = NULL, dubious metacells detection will be performed on the metacell partition represented by the first column of cell_membership. Default is NULL.</li>
<li>
<code>assay_type</code>: A charater of the type of data assay you are using (‘RNA’ or ‘ATAC’), depending on which different normalization would be used. Default is ‘RNA’.</li>
<li>
<code>feature_use</code>: An integer of the number of genes to use in metacell testing. Default is 2000.</li>
<li>
<code>aggregate_method</code>: A character of the method to aggregate single cell profiles into metacell profiles. Default is ‘mean’, which means the metacell expression is computed as the mean of the single cell expression within.</li>
<li>
<code>Nrep</code>: An integer of the number of permutation repetitions we use when deriving the null. Default is 1.</li>
<li>
<code>test_cutoff</code>: A float of the test size for dubious metacell detection testing. Default is 0.05.</li>
<li>
<code>thre_smooth</code>: A boolean indicating whether to smooth the threshold function (lowess smoothing). Default is TRUE.</li>
<li>
<code>thre_bw</code>: A float. If thre_smooth is True, thre_bw specifies the bandwidth for smoothing. Default is 1/6.</li>
</ul>
<p>The mcRigor detection results, indicating whether each metacell is classified as dubious or trustworthy, are stored in the <code>mc_res</code> field of the output object <code>detect_res</code>.</p>
<p>We implement the second functionality: <strong>optimizing the hyperparameter</strong> of a metacell partitioning method. We use SEACells as an example here.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">optimize_res</span> <span class="op">=</span> <span class="fu"><a href="reference/mcRigor_OPTIMIZE.html">mcRigor_OPTIMIZE</a></span><span class="op">(</span>obj_singlecell <span class="op">=</span> <span class="va">obj_singlecell</span>, cell_membership <span class="op">=</span> <span class="va">cell_membership_all</span><span class="op">)</span></span></code></pre></div>
<p>Besides the arguments of <code><a href="reference/mcRigor_DETECT.html">mcRigor_DETECT()</a></code>, <code><a href="reference/mcRigor_OPTIMIZE.html">mcRigor_OPTIMIZE()</a></code> has one additional argument:</p>
<ul>
<li>
<code>Gammas</code>: A vector of integers or characters to specify the candidate pool of granularity levels to consider in optimization. Default is NULL, which means all the granularity levels in <code>cell_membership</code> will be considered.</li>
<li>
<code>weight</code>: A float between 0 and 1, which sepcifies the weight for dubious rate in the tradeoff for metacell partition evaluation. Default is 0.5.</li>
</ul>
<p>The output <code>optimize_res</code> contains the optimal granularity level (<code>best_granularity_level</code>) and the Seurat object of metacells generated under the the optimal granularity level (<code>opt_metacell</code>). We can visualize the optimal metacell partition using function <code>mcRigor_projection</code>, with the metacells projected to the two-dimensional embedding space of single cells and the detected dubious metacells marked by red circles</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">opt_metacell</span> <span class="op">=</span> <span class="va">optimize_res</span><span class="op">$</span><span class="va">opt_metacell</span></span>
<span></span>
<span><span class="va">sc_membership</span> <span class="op">=</span> <span class="va">opt_metacell</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">cell_membership</span><span class="op">$</span><span class="va">Metacell</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sc_membership</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">opt_metacell</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">cell_membership</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span> <span class="op">=</span> <span class="fu"><a href="reference/mcRigor_projection.html">mcRigor_projection</a></span><span class="op">(</span>obj_singlecell <span class="op">=</span> <span class="va">obj_singlecell</span>, sc_membership <span class="op">=</span> <span class="va">sc_membership</span>,</span>
<span>                           color_field <span class="op">=</span> <span class="st">'celltype.l1'</span>,</span>
<span>                           dub_mc_test.label <span class="op">=</span> <span class="cn">T</span>, test_stats <span class="op">=</span> <span class="va">optimize_res</span><span class="op">$</span><span class="va">TabMC</span>, Thre <span class="op">=</span> <span class="va">optimize_res</span><span class="op">$</span><span class="va">thre</span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="reference/figures/mcRigor_projection.png" width="600"></p>
</div>
<div class="section level2">
<h2 id="tutorials">Tutorials<a name="tutorials"></a>
<a class="anchor" aria-label="anchor" href="#tutorials"></a>
</h2>
<p>For all detailed tutorials, please check the <a href="https://jsb-ucla.github.io/mcRigor/index.html" class="external-link">website</a>. The tutorials will demonstrate the two main functionalities of <strong>mcRigor</strong> on a semi-synthetic dataset of bone marrow mononuclear cells measured by CITE-seq.</p>
<ul>
<li><a href="https://jsb-ucla.github.io/mcRigor/articles/mcRigor-1-detect-dubmc.html" class="external-link">Functionality 1: detect dubious metacells for a given metacell partition</a></li>
<li><a href="https://jsb-ucla.github.io/mcRigor/articles/mcRigor-2-optimize.html" class="external-link">Functionality 2: optimize metacell partitioning</a></li>
</ul>
</div>
<div class="section level2">
<h2 id="related-paper">Related Paper<a name="related-manuscripts"></a>
<a class="anchor" aria-label="anchor" href="#related-paper"></a>
</h2>
<ul>
<li>
<strong>mcRigor</strong>: <a href="https://doi.org/10.1101/2024.10.30.621093" class="external-link">Liu, P. and Li, J.J. mcRigor: a statistical method to enhance the rigor of metacell partitioning in single-cell data analysis. <em>bioRxiv</em> (2024).</a>
</li>
</ul>
</div>
</div>
  </main><aside class="col-md-3"><div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li>
<a href="https://opensource.org/licenses/mit-license.php" class="external-link">MIT</a> + file <a href="LICENSE-text.html">LICENSE</a>
</li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing mcRigor</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Pan Liu <br><small class="roles"> Maintainer </small>   </li>
</ul>
</div>



  </aside>
</div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pan Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
