---
title: "Extension: mcRigor two-step"
author: "Pan Liu"
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('mcRigor')`"
vignette: >
  %\VignetteIndexEntry{mcRigor-4-twostep}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
tools::R_user_dir("mcRigor", which="cache")
```

```{r setup, message=FALSE, warning=FALSE, results='hide'}
#tools::R_user_dir("mcRigor", which="cache")
library(mcRigor)
library(Seurat)
library(ggplot2)
```

## Introduction

### Overview

In this tutorial, we demonstrate how to use the **mcRigor two-step** extension of **mcRigor** to further dissect previously identified dubious metacells and reorganize their constituent cells into more trustworthy metacells. We will apply **mcRigor two-step** to a semi-synthetic single-cell RNA sequencing (scRNA-seq) dataset with known ground-truth metacell trustworthiness.

As its name suggests, **mcRigor two-step** operates in two consecutive steps:

1.  **Step 1** (function `mcRigorTS_Step1`): Identify the single cells that constitute dubious metacells and need to be re-partitioned.

2.  **Step 2** (function `mcRigorTS_Step2`): Re-partition the identified single cells under the optimized granularity level and output the final metacell partition.

## Input preparation

As for the main functionalities of mcRigor, we also work on two main inputs in mcRigor two-step: 1. the raw scRNA-seq data and 2. candidate metacell partitions generated by either existing metacell partitioning methods or ad hoc approaches. The raw scRNA-seq data needs to be provided as a Seurat object, `obj_singlecell`. The semi-synthetic scRNA-seq data, whose generation process is described in [Liu and Li, 2024](https://www.biorxiv.org/content/10.1101/2024.10.30.621093v2), stored as a rds file `syn.rds`, is available with the mcRigor package as an example. The metacell partitions should be provided as a dataframe, `cell_membership`, showing the assignment of single cells to metacells in each partition. Specifically, each column of this dataframe should represent the matacell partition corresponding to one granularity level and each row of the dataframe should represent one single cell. Note that we require the column namse of the dataframe to be set as the granularity level values in the character type. The metacell partitions for the semi-synthetic scRNA-seq data generated by the SEACells method ([Persad et al., 2023](https://www.nature.com/articles/s41587-023-01716-9)), stored as a csv file `seacells_cell_membership_rna_syn.csv`, is available with the mcRigor package as an example. This csv file contains series of metacell partitions, which were generated under different granularity levels. Note that granularity level, $\gamma$, is a key parameter for metacell partitioning and is defined as the ratio of the number of single cells to the number of metacells. We first load these inputs:

```{r}
sc_dir = system.file('extdata', 'syn.rds', package = 'mcRigor')
obj_singlecell= readRDS(file = sc_dir)

membership_dir = system.file('extdata', 'seacells_cell_membership_rna_syn.csv', package = 'mcRigor')
cell_membership <- read.csv(file = membership_dir, check.names = F, row.names = 1)
```

## Step 1: identify single cells that constitute the dubious metacells

In this tutorial, we focus on the metacell partition corresponding to the granularity level $\gamma = 50$, specified by `tgamma` = 50. We first run the main mcRigor function `mcRigor_DETECT` to check whether this partition contains any dubious metacells. If no dubious metacells are detected, there is no need to apply mcRigor two-step.

```{r, results='hide'}
tgamma = 50
detect_res = mcRigor_DETECT(obj_singlecell = obj_singlecell, cell_membership = cell_membership, tgamma = tgamma)
```
```{r}
table(detect_res$mc_res)
```

We proceed with mcRigor two-step because dubious metacells exist in this partition We call the function `mcRigorTS_Step1` to identify the single cells that constitute dubious metacells (under a lower divergence score threshold) and need to be re-partitioned. Note that we input the permutation results, `detect_res$TabMC`, obtained from the previous code block to save computation time here, although it is also acceptable to leave this argument as its default value (`NULL`). We need to specify the metacell partitioning method (e.g., SEACells, MetaCell, MetaCell2, SuperCell, or MetaQ) by setting the argument `method`, ensuring that the function generates outputs in the correct format.

```{r}
sc_membership = cell_membership[[as.character(tgamma)]]
names(sc_membership) = rownames(cell_membership)

step1_res = mcRigorTS_Step1(obj_singlecell = obj_singlecell, sc_membership = sc_membership, TabMC = detect_res$TabMC, method = 'seacells')
```

The Seurat object of single cells that need to be re-partitioned is stored in the `obj_sc_dub` field of the output `step1_res`. If the metacell partitioning method used is SEACells, MetaCell2, or MetaQ,, `mcRigorTS_Step1` will automatically generate two CSV files (`counts.csv` and `metadata.csv`) in the current working directory. These files can be directly used as input for the corresponding metacell partitioning methods in Python.

```{r}
step1_res$obj_sc_dub
list.files("seacells/")
```

## Step 2: Re-organize the single cells into more trustworthy metacells

We then need to apply the chosen metacell partitioning method (e.g., SEACells, MetaCell, MetaCell2, SuperCell, MetaQ) to these identified single cells to obtain more trustworthy metacell partitions for them, under granularity level smaller than `tgamma` (recommend obtaining all the candidate paritions corresponding to granularity levels from 2 to `tgamma`-1). One can refer to the [Implementing metacell partitioning methods](https://jsb-ucla.github.io/mcRigor/articles/mcRigor-3-mc-method.html) tutorial for more guidnance on how to obtain the metacell partitions. The candidate metacell partitions can be stored in csv file and read into an R object `cell_membership_twostep` using `cell_membership_twostep <- read.csv(file = "/path/to/the/csv/file", check.names = F, row.names = 1)`. For illustration purpose in this tutorial, we set `cell_membership_twostep` as the partitions corresponding to small granularity levels in `cell_membership_all`.

```{r}
cell_membership_twostep = cell_membership[ , as.numeric(names(cell_membership)) < 50]
cell_membership_twostep = cell_membership_twostep[colnames(step1_res$obj_sc_dub),]
```

We call the function `mcRigorTS_Step2` to select the optimal granularity level and output the final metacell partition with remaining dubious metacells flagged.

```{r, results='hide'}
step2_res = mcRigorTS_Step2(step1_res = step1_res, obj_singlecell = obj_singlecell, cell_membership_twostep = cell_membership_twostep)
```

The Seurat object of the final metacells are stored in the `obj_metacell_final` field of the output `step2_res` with the dubious metacell detection results recorded in its metadata under the name `mcRigor`. Another field, `obj_metacell_final_withsc`, is the Seurat object of the final metacells where all dubious metacells have been dissected into their constituent single cells (i.e., no dubious metacells remain). Users can choose which object to use depending on their analytical needs. In addtion, the output `step2_res` includes two visualization fields: `plot` and `plot_withsc`, which correspond to UMAP plots of single cells with `obj_metacell_final` and `obj_metacell_final_withsc` metacells projected onto the UMAP space, respectively.

```{r}
step2_res$obj_metacell_final
step2_res$plot
```


## Session information

```{r}
sessionInfo()
```
