---
title: "Functionality 1: detect dubious metacells for a given metacell partition"
author: "Pan Liu"
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('mcRigor')`"
vignette: >
  %\VignetteIndexEntry{mcRigor-detect-dubmc}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
tools::R_user_dir("mcRigor", which="cache")
```

```{r setup, message=FALSE, warning=FALSE, results='hide'}
#tools::R_user_dir("mcRigor", which="cache")
library(mcRigor)
library(Seurat)
library(ggplot2)
```

## Introduction

In this tutorial, we will show how to use mcRigor to detect dubious metacells for a given metacell partition. We will demonstrate this functionality of mcRigor on a semi-synthetic single cell RNA sequencing (scRNA-seq) dataset with known ground truth trustworthiness of metacells.

## Input preparation

Two main inputs are required for this functionality: 1. the raw scRNA-seq data and 2. a given metacell partition generated by either existing metacell partitioning methods or ad hoc approaches. The raw scRNA-seq data needs to be provided as a Seurat object, `obj_singlecell`. The semi-synthetic scRNA-seq data, whose generation process is described in [Liu and Li, 2024](https://www.biorxiv.org/content/10.1101/2024.10.30.621093v2), stored as a rds file `syn.rds`, is available with the mcRigor package as an example. We first load the data.
```{r}
sc_dir = system.file('extdata', 'syn.rds', package = 'mcRigor')
obj_singlecell= readRDS(file = sc_dir)
obj_singlecell
```
The metacell partition should be provided as a dataframe, `cell_membership`, showing the assignment of single cells to metacells. Specifically, this dataframe has only one column and each row of the dataframe represents one single cell. The metacell partitions for the semi-synthetic scRNA-seq data generated by the SEACells method ([Persad et al., 2023](https://www.nature.com/articles/s41587-023-01716-9)), stored as a csv file `seacells_cell_membership_rna_syn.csv`, is available with the mcRigor package as an example. This csv file contains series of metacell partitions, which were generated under different granularity levels. Note that granularity level, $\gamma$, is a key parameter for metacell partitioning and is defined as the ratio of the number of single cells to the number of metacells. In this tutorial, we focus on the metacell partition given by granularity level $\gamma = 50$. 
```{r}
membership_dir = system.file('extdata', 'seacells_cell_membership_rna_syn.csv', package = 'mcRigor')
cell_membership_all <- read.csv(file = membership_dir, check.names = F, row.names = 1)
cell_membership <- cell_membership_all['50']
head(cell_membership)
```


## Detection of dubious metacells

We call the function `mcRigor_DETECT` to detect dubious metacells for the metacell partition represented by `cell_membership`.
```{r, results='hide'}
detect_res = mcRigor_DETECT(obj_singlecell = obj_singlecell, cell_membership = cell_membership)
```
The Seurat object of metacells are stored in the `obj_metacell` field of the output `detect_res`.The mcRigor detection results are recorded in the `mc_res` field of the output `detect_res` as well as the metadata of the Seurat object with name `mcRigor`.
```{r}
table(detect_res$mc_res)
obj_metacell = detect_res$obj_metacell
head(obj_metacell$mcRigor)
```

## Visualization

The function `mcRigor_projection` can draw the metacells projected to the two-dimensional embedding space of single cells and mark the detected dubious metacells
```{r, results='hide', fig.keep='all'}
sc_membership = obj_metacell@misc$cell_membership$Metacell
names(sc_membership) = rownames(obj_metacell@misc$cell_membership)

plot = mcRigor_projection(obj_singlecell = obj_singlecell, sc_membership = sc_membership,
                           color_field = 'celltype.l1',
                           dub_mc_test.label = T, test_stats = detect_res$TabMC, Thre = detect_res$thre)
plot
```

The dubious metacells are marked by red circles while the trustworthy metacells are with black circles.



## Session information
```{r}
sessionInfo()
```
