---
title: 'Functionality 2: optimize metacell partitioning'
author: "Pan Liu"
output: 
  BiocStyle::html_document:
    self_contained: yes
    toc: true
    toc_float: true
    toc_depth: 2
    code_folding: show
date: "`r doc_date()`"
package: "`r pkg_ver('mcRigor')`"
vignette: >
  %\VignetteIndexEntry{mcRigor-detect-dubmc}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}  
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
tools::R_user_dir("mcRigor", which="cache")
```

```{r setup, message=FALSE, warning=FALSE, results='hide'}
#tools::R_user_dir("mcRigor", which="cache")
library(mcRigor)
library(Seurat)
library(ggplot2)
```

## Introduction

In this tutorial, we will show how to use mcRigor to evaluate each metacell partition and select the **optimal granularity level** for a specific metacell partitioning method. Note that granularity level, $\gamma$, is a key parameter for metacell partitioning defined as the ratio of the number of single cells to the number of metacells, and decent selection of it is of vital importance to ensure the unbiasedness of the metacell profiles and the reliability fo the downstream analysis. We will demonstrate this functionality of mcRigor on a semi-synthetic single cell RNA sequencing (scRNA-seq) dataset with known trustworthiness of metacells and optimal metacell partition.

## Input preparation

Two main inputs are required for this functionality: 1. the raw scRNA-seq data and 2. a series of candidate metacell partitions generated by a specific metacell partitioning method with a series of candidate granularity levels. The raw scRNA-seq data needs to be provided as a Seurat object, `obj_singlecell`. The semi-synthetic scRNA-seq data, whose generation process is described in [Liu and Li, 2024](https://www.biorxiv.org/content/10.1101/2024.10.30.621093v2), stored as a rds file `syn.rds`, is available with the mcRigor package as an example. We first load the data.
```{r}
sc_dir = system.file('extdata', 'syn.rds', package = 'mcRigor')
obj_singlecell= readRDS(file = sc_dir)
obj_singlecell
```
The candidate metacell partitions should be provided as a dataframe, `cell_membership`, showing the assignment of single cells to metacells in each partition. Specifically, each column of this dataframe should represent the matacell partition corresponding to one granularity level and each row of the dataframe should represent one single cell. Note that we require the column namse of the dataframe to be set as the granularity level values in the character type. The metacell partitions for the semi-synthetic scRNA-seq data generated by the SEACells method ([Persad et al., 2023](https://www.nature.com/articles/s41587-023-01716-9)), stored as a csv file `seacells_cell_membership_rna_syn.csv`, is available with the mcRigor package as an example.
```{r}
membership_dir = system.file('extdata', 'seacells_cell_membership_rna_syn.csv', package = 'mcRigor')
cell_membership <- read.csv(file = membership_dir, check.names = F, row.names = 1)
cell_membership[1:3,1:3]
```


## Optimization of hyperparameter selection

We call the function `mcRigor_OPTIMIZE` to evaluate each metacell partition in `cell_membership` and select the optimal one among them.
```{r, results='hide'}
optimize_res = mcRigor_OPTIMIZE(obj_singlecell = obj_singlecell, cell_membership = cell_membership)
```
The evaluation scores for the provided metacell partitions are stored in the `score` field of the output `optimize_res`, and we can draw the line plot of the evaluation scores.
```{r}
head(optimize_res$scores)
optimize_res$optim_plot
```
The output `optimize_res` contains the optimal granularity level (`best_granularity_level`), the evaluation score for the metacell partition given by the optimal granularity level selected (`best_score`), and the Seurat object of metacells generated under the the optimal granularity level (`opt_metacell`).
```{r}
opt_metacell = optimize_res$opt_metacell
opt_metacell
```
Note the optimal metacell partition may still contain dubious metacells. The mcRigor dubious metacell detection results are recorded in the metadata of `opt_metacell` with name `mcRigor`. The user may choose to further exclude the dubious metacells from the optimal partition if lost of information is bearable. 
```{r}
opt_metacell_tuned = subset(opt_metacell, mcRigor == 'trustworthy')
```


## Visualization

The function `mcRigor_projection` can visualize the optimal metacell partition, with the metacells projected to the two-dimensional embedding space of single cells and mark the detected dubious metacells
```{r, results='hide', fig.keep='all'}
sc_membership = opt_metacell@misc$cell_membership$Metacell
names(sc_membership) = rownames(opt_metacell@misc$cell_membership)

plot = mcRigor_projection(obj_singlecell = obj_singlecell, sc_membership = sc_membership,
                           color_field = 'celltype.l1',
                           dub_mc_test.label = T, test_stats = optimize_res$TabMC, Thre = optimize_res$thre)
plot
```

The dubious metacells are marked by red circles while the trustworthy metacells are with black circles.



## Session information
```{r}
sessionInfo()
```
