<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Functionality 2: optimize metacell partitioning • mcRigor</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Functionality 2: optimize metacell partitioning">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">mcRigor</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/mcRigor-1-detect-dubmc.html">Functionality 1: detect dubious metacells for a given metacell partition</a></li>
    <li><a class="dropdown-item" href="../articles/mcRigor-2-optimize.html">Functionality 2: optimize metacell partitioning</a></li>
    <li><a class="dropdown-item" href="../articles/mcRigor-3-mc-method.html">Implementing metacell partitioning methods</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Functionality 2: optimize metacell partitioning</h1>
                        <h4 data-toc-skip class="author">Pan Liu</h4>
            
            <h4 data-toc-skip class="date">26 November 2024</h4>
      

      <div class="d-none name"><code>mcRigor-2-optimize.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#tools::R_user_dir("mcRigor", which="cache")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jsb-ucla.github.io/mcRigor/">mcRigor</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this tutorial, we will show how to use mcRigor to evaluate each
metacell partition and select the <strong>optimal granularity
level</strong> for a specific metacell partitioning method. Note that
granularity level,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>γ</mi><annotation encoding="application/x-tex">\gamma</annotation></semantics></math>,
is a key parameter for metacell partitioning defined as the ratio of the
number of single cells to the number of metacells, and decent selection
of it is of vital importance to ensure the unbiasedness of the metacell
profiles and the reliability fo the downstream analysis. We will
demonstrate this functionality of mcRigor on a semi-synthetic single
cell RNA sequencing (scRNA-seq) dataset with known trustworthiness of
metacells and optimal metacell partition.</p>
</div>
<div class="section level2">
<h2 id="input-preparation">Input preparation<a class="anchor" aria-label="anchor" href="#input-preparation"></a>
</h2>
<p>Two main inputs are required for this functionality: 1. the raw
scRNA-seq data and 2. a series of candidate metacell partitions
generated by a specific metacell partitioning method with a series of
candidate granularity levels. The raw scRNA-seq data needs to be
provided as a Seurat object, <code>obj_singlecell</code>. The
semi-synthetic scRNA-seq data, whose generation process is described in
<a href="https://www.biorxiv.org/content/10.1101/2024.10.30.621093v2" class="external-link">Liu
and Li, 2024</a>, stored as a rds file <code>syn.rds</code>, is
available with the mcRigor package as an example. We first load the
data.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'syn.rds'</span>, package <span class="op">=</span> <span class="st">'mcRigor'</span><span class="op">)</span></span>
<span><span class="va">obj_singlecell</span><span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">sc_dir</span><span class="op">)</span></span>
<span><span class="va">obj_singlecell</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 13400 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 2000 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span></code></pre></div>
<p>The candidate metacell partitions should be provided as a dataframe,
<code>cell_membership</code>, showing the assignment of single cells to
metacells in each partition. Specifically, each column of this dataframe
should represent the matacell partition corresponding to one granularity
level and each row of the dataframe should represent one single cell.
Note that we require the column namse of the dataframe to be set as the
granularity level values in the character type. The metacell partitions
for the semi-synthetic scRNA-seq data generated by the SEACells method
(<a href="https://www.nature.com/articles/s41587-023-01716-9" class="external-link">Persad et
al., 2023</a>), stored as a csv file
<code>seacells_cell_membership_rna_syn.csv</code>, is available with the
mcRigor package as an example.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">membership_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'seacells_cell_membership_rna_syn.csv'</span>, package <span class="op">=</span> <span class="st">'mcRigor'</span><span class="op">)</span></span>
<span><span class="va">cell_membership</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">membership_dir</span>, check.names <span class="op">=</span> <span class="cn">F</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">cell_membership</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">3</span>,<span class="fl">1</span><span class="op">:</span><span class="fl">3</span><span class="op">]</span></span>
<span><span class="co">#&gt;                               100                       99</span></span>
<span><span class="co">#&gt; 1_Cell1 mc100-allcells-SEACell-86 mc99-allcells-SEACell-35</span></span>
<span><span class="co">#&gt; 2_Cell1 mc100-allcells-SEACell-26 mc99-allcells-SEACell-35</span></span>
<span><span class="co">#&gt; 3_Cell1 mc100-allcells-SEACell-26 mc99-allcells-SEACell-35</span></span>
<span><span class="co">#&gt;                                98</span></span>
<span><span class="co">#&gt; 1_Cell1 mc98-allcells-SEACell-131</span></span>
<span><span class="co">#&gt; 2_Cell1 mc98-allcells-SEACell-131</span></span>
<span><span class="co">#&gt; 3_Cell1 mc98-allcells-SEACell-131</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="optimization-of-hyperparameter-selection">Optimization of hyperparameter selection<a class="anchor" aria-label="anchor" href="#optimization-of-hyperparameter-selection"></a>
</h2>
<p>We call the function <code>mcRigor_OPTIMIZE</code> to evaluate each
metacell partition in <code>cell_membership</code> and select the
optimal one among them.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">optimize_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/mcRigor_OPTIMIZE.html">mcRigor_OPTIMIZE</a></span><span class="op">(</span>obj_singlecell <span class="op">=</span> <span class="va">obj_singlecell</span>, cell_membership <span class="op">=</span> <span class="va">cell_membership</span><span class="op">)</span></span></code></pre></div>
<p>The evaluation scores for the provided metacell partitions are stored
in the <code>score</code> field of the output <code>optimize_res</code>,
and we can draw the line plot of the evaluation scores.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">optimize_res</span><span class="op">$</span><span class="va">scores</span><span class="op">)</span></span>
<span><span class="co">#&gt;   gamma    DubRate  ZeroRate     Score</span></span>
<span><span class="co">#&gt; 1     2 0.01106557 0.9201214 0.5344065</span></span>
<span><span class="co">#&gt; 2     3 0.01193333 0.9021871 0.5429398</span></span>
<span><span class="co">#&gt; 3     4 0.01272903 0.8864460 0.5504125</span></span>
<span><span class="co">#&gt; 4     6 0.01418433 0.8618650 0.5619753</span></span>
<span><span class="co">#&gt; 5     7 0.01459026 0.8507913 0.5673092</span></span>
<span><span class="co">#&gt; 6     8 0.01466195 0.8412219 0.5720581</span></span>
<span><span class="va">optimize_res</span><span class="op">$</span><span class="va">optim_plot</span></span></code></pre></div>
<p><img src="mcRigor-2-optimize_files/figure-html/unnamed-chunk-5-1.png" width="700"></p>
<p>The output <code>optimize_res</code> contains the optimal granularity
level (<code>best_granularity_level</code>), the evaluation score for
the metacell partition given by the optimal granularity level selected
(<code>best_score</code>), and the Seurat object of metacells generated
under the the optimal granularity level (<code>opt_metacell</code>).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">opt_metacell</span> <span class="op">=</span> <span class="va">optimize_res</span><span class="op">$</span><span class="va">opt_metacell</span></span>
<span><span class="va">opt_metacell</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 319 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span></code></pre></div>
<p>Note the optimal metacell partition may still contain dubious
metacells. The mcRigor dubious metacell detection results are recorded
in the metadata of <code>opt_metacell</code> with name
<code>mcRigor</code>. The user may choose to further exclude the dubious
metacells from the optimal partition if lost of information is
bearable.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">opt_metacell_tuned</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/subset.html" class="external-link">subset</a></span><span class="op">(</span><span class="va">opt_metacell</span>, <span class="va">mcRigor</span> <span class="op">==</span> <span class="st">'trustworthy'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="visualization">Visualization<a class="anchor" aria-label="anchor" href="#visualization"></a>
</h2>
<p>The function <code>mcRigor_projection</code> can visualize the
optimal metacell partition, with the metacells projected to the
two-dimensional embedding space of single cells and mark the detected
dubious metacells</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_membership</span> <span class="op">=</span> <span class="va">opt_metacell</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">cell_membership</span><span class="op">$</span><span class="va">Metacell</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sc_membership</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">opt_metacell</span><span class="op">@</span><span class="va">misc</span><span class="op">$</span><span class="va">cell_membership</span><span class="op">)</span></span>
<span></span>
<span><span class="va">plot</span> <span class="op">=</span> <span class="fu"><a href="../reference/mcRigor_projection.html">mcRigor_projection</a></span><span class="op">(</span>obj_singlecell <span class="op">=</span> <span class="va">obj_singlecell</span>, sc_membership <span class="op">=</span> <span class="va">sc_membership</span>,</span>
<span>                           color_field <span class="op">=</span> <span class="st">'celltype.l1'</span>,</span>
<span>                           dub_mc_test.label <span class="op">=</span> <span class="cn">T</span>, test_stats <span class="op">=</span> <span class="va">optimize_res</span><span class="op">$</span><span class="va">TabMC</span>, Thre <span class="op">=</span> <span class="va">optimize_res</span><span class="op">$</span><span class="va">thre</span><span class="op">)</span></span>
<span><span class="va">plot</span></span></code></pre></div>
<p><img src="mcRigor-2-optimize_files/figure-html/unnamed-chunk-8-1.png" width="700"></p>
<p>The dubious metacells are marked by red circles while the trustworthy
metacells are with black circles.</p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.4.2 (2024-10-31)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 22.04.5 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.20.so;  LAPACK version 3.10.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_3.5.1      Seurat_5.1.0       SeuratObject_5.0.2 sp_2.1-4          </span></span>
<span><span class="co">#&gt; [5] mcRigor_1.0        BiocStyle_2.34.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] RColorBrewer_1.1-3     jsonlite_1.8.9         magrittr_2.0.3        </span></span>
<span><span class="co">#&gt;   [4] spatstat.utils_3.1-1   farver_2.1.2           rmarkdown_2.29        </span></span>
<span><span class="co">#&gt;   [7] fs_1.6.5               ragg_1.3.3             vctrs_0.6.5           </span></span>
<span><span class="co">#&gt;  [10] ROCR_1.0-11            spatstat.explore_3.3-3 htmltools_0.5.8.1     </span></span>
<span><span class="co">#&gt;  [13] sass_0.4.9             sctransform_0.4.1      parallelly_1.39.0     </span></span>
<span><span class="co">#&gt;  [16] KernSmooth_2.23-24     bslib_0.8.0            htmlwidgets_1.6.4     </span></span>
<span><span class="co">#&gt;  [19] desc_1.4.3             ica_1.0-3              plyr_1.8.9            </span></span>
<span><span class="co">#&gt;  [22] plotly_4.10.4          zoo_1.8-12             cachem_1.1.0          </span></span>
<span><span class="co">#&gt;  [25] igraph_2.1.1           mime_0.12              lifecycle_1.0.4       </span></span>
<span><span class="co">#&gt;  [28] pkgconfig_2.0.3        Matrix_1.7-1           R6_2.5.1              </span></span>
<span><span class="co">#&gt;  [31] fastmap_1.2.0          fitdistrplus_1.2-1     future_1.34.0         </span></span>
<span><span class="co">#&gt;  [34] shiny_1.9.1            digest_0.6.37          colorspace_2.1-1      </span></span>
<span><span class="co">#&gt;  [37] patchwork_1.3.0        tensor_1.5             RSpectra_0.16-2       </span></span>
<span><span class="co">#&gt;  [40] irlba_2.3.5.1          textshaping_0.4.0      labeling_0.4.3        </span></span>
<span><span class="co">#&gt;  [43] progressr_0.15.1       fansi_1.0.6            spatstat.sparse_3.1-0 </span></span>
<span><span class="co">#&gt;  [46] httr_1.4.7             polyclip_1.10-7        abind_1.4-8           </span></span>
<span><span class="co">#&gt;  [49] compiler_4.4.2         withr_3.0.2            fastDummies_1.7.4     </span></span>
<span><span class="co">#&gt;  [52] MASS_7.3-61            tools_4.4.2            lmtest_0.9-40         </span></span>
<span><span class="co">#&gt;  [55] httpuv_1.6.15          future.apply_1.11.3    goftest_1.2-3         </span></span>
<span><span class="co">#&gt;  [58] glue_1.8.0             nlme_3.1-166           promises_1.3.0        </span></span>
<span><span class="co">#&gt;  [61] grid_4.4.2             Rtsne_0.17             cluster_2.1.6         </span></span>
<span><span class="co">#&gt;  [64] reshape2_1.4.4         generics_0.1.3         gtable_0.3.6          </span></span>
<span><span class="co">#&gt;  [67] spatstat.data_3.1-4    tidyr_1.3.1            data.table_1.16.2     </span></span>
<span><span class="co">#&gt;  [70] utf8_1.2.4             spatstat.geom_3.3-4    RcppAnnoy_0.0.22      </span></span>
<span><span class="co">#&gt;  [73] ggrepel_0.9.6          RANN_2.6.2             pillar_1.9.0          </span></span>
<span><span class="co">#&gt;  [76] stringr_1.5.1          spam_2.11-0            RcppHNSW_0.6.0        </span></span>
<span><span class="co">#&gt;  [79] later_1.3.2            splines_4.4.2          dplyr_1.1.4           </span></span>
<span><span class="co">#&gt;  [82] lattice_0.22-6         survival_3.7-0         deldir_2.0-4          </span></span>
<span><span class="co">#&gt;  [85] tidyselect_1.2.1       miniUI_0.1.1.1         pbapply_1.7-2         </span></span>
<span><span class="co">#&gt;  [88] knitr_1.49             gridExtra_2.3          bookdown_0.41         </span></span>
<span><span class="co">#&gt;  [91] scattermore_1.2        xfun_0.49              matrixStats_1.4.1     </span></span>
<span><span class="co">#&gt;  [94] stringi_1.8.4          lazyeval_0.2.2         yaml_2.3.10           </span></span>
<span><span class="co">#&gt;  [97] evaluate_1.0.1         codetools_0.2-20       tibble_3.2.1          </span></span>
<span><span class="co">#&gt; [100] BiocManager_1.30.25    cli_3.6.3              uwot_0.2.2            </span></span>
<span><span class="co">#&gt; [103] xtable_1.8-4           reticulate_1.40.0      systemfonts_1.1.0     </span></span>
<span><span class="co">#&gt; [106] munsell_0.5.1          jquerylib_0.1.4        Rcpp_1.0.13-1         </span></span>
<span><span class="co">#&gt; [109] globals_0.16.3         spatstat.random_3.3-2  png_0.1-8             </span></span>
<span><span class="co">#&gt; [112] spatstat.univar_3.1-1  parallel_4.4.2         pkgdown_2.1.1         </span></span>
<span><span class="co">#&gt; [115] dotCall64_1.2          listenv_0.9.1          viridisLite_0.4.2     </span></span>
<span><span class="co">#&gt; [118] scales_1.3.0           ggridges_0.5.6         leiden_0.4.3.1        </span></span>
<span><span class="co">#&gt; [121] purrr_1.0.2            rlang_1.1.4            cowplot_1.1.3</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pan Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.</p>
</div>

    </footer>
</div>





  </body>
</html>
