<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Extension: mcRigor two-step • mcRigor</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Extension: mcRigor two-step">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">mcRigor</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/mcRigor-1-detect-dubmc.html">Functionality 1: detect dubious metacells for a given metacell partition</a></li>
    <li><a class="dropdown-item" href="../articles/mcRigor-2-optimize.html">Functionality 2: optimize metacell partitioning</a></li>
    <li><a class="dropdown-item" href="../articles/mcRigor-3-mc-method.html">Implementing metacell partitioning methods</a></li>
    <li><a class="dropdown-item" href="../articles/mcRigor-4-twostep.html">Extension: mcRigor two-step</a></li>
  </ul>
</li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">

      <h1>Extension: mcRigor two-step</h1>
                        <h4 data-toc-skip class="author">Pan Liu</h4>
            
            <h4 data-toc-skip class="date">18 October 2025</h4>
      

      <div class="d-none name"><code>mcRigor-4-twostep.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co">#tools::R_user_dir("mcRigor", which="cache")</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://jsb-ucla.github.io/mcRigor/">mcRigor</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://satijalab.org/seurat" class="external-link">Seurat</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<div class="section level3">
<h3 id="overview">Overview<a class="anchor" aria-label="anchor" href="#overview"></a>
</h3>
<p>In this tutorial, we demonstrate how to use the <strong>mcRigor
two-step</strong> extension of <strong>mcRigor</strong> to further
dissect previously identified dubious metacells and reorganize their
constituent cells into more trustworthy metacells. We will apply
<strong>mcRigor two-step</strong> to a semi-synthetic single-cell RNA
sequencing (scRNA-seq) dataset with known ground-truth metacell
trustworthiness.</p>
<p>As its name suggests, <strong>mcRigor two-step</strong> operates in
two consecutive steps:</p>
<ol style="list-style-type: decimal">
<li><p><strong>Step 1</strong> (function <code>mcRigorTS_Step1</code>):
Identify the single cells that constitute dubious metacells and need to
be re-partitioned.</p></li>
<li><p><strong>Step 2</strong> (function <code>mcRigorTS_Step2</code>):
Re-partition the identified single cells under the optimized granularity
level and output the final metacell partition.</p></li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="input-preparation">Input preparation<a class="anchor" aria-label="anchor" href="#input-preparation"></a>
</h2>
<p>As for the main functionalities of mcRigor, we also work on two main
inputs in mcRigor two-step: 1. the raw scRNA-seq data and 2. candidate
metacell partitions generated by either existing metacell partitioning
methods or ad hoc approaches. The raw scRNA-seq data needs to be
provided as a Seurat object, <code>obj_singlecell</code>. The
semi-synthetic scRNA-seq data, whose generation process is described in
<a href="https://www.biorxiv.org/content/10.1101/2024.10.30.621093v2" class="external-link">Liu
and Li, 2024</a>, stored as a rds file <code>syn.rds</code>, is
available with the mcRigor package as an example. The metacell
partitions should be provided as a dataframe,
<code>cell_membership</code>, showing the assignment of single cells to
metacells in each partition. Specifically, each column of this dataframe
should represent the matacell partition corresponding to one granularity
level and each row of the dataframe should represent one single cell.
Note that we require the column namse of the dataframe to be set as the
granularity level values in the character type. The metacell partitions
for the semi-synthetic scRNA-seq data generated by the SEACells method
(<a href="https://www.nature.com/articles/s41587-023-01716-9" class="external-link">Persad et
al., 2023</a>), stored as a csv file
<code>seacells_cell_membership_rna_syn.csv</code>, is available with the
mcRigor package as an example. This csv file contains series of metacell
partitions, which were generated under different granularity levels.
Note that granularity level,
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mi>γ</mi><annotation encoding="application/x-tex">\gamma</annotation></semantics></math>,
is a key parameter for metacell partitioning and is defined as the ratio
of the number of single cells to the number of metacells. We first load
these inputs:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'syn.rds'</span>, package <span class="op">=</span> <span class="st">'mcRigor'</span><span class="op">)</span></span>
<span><span class="va">obj_singlecell</span><span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">readRDS</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">sc_dir</span><span class="op">)</span></span>
<span></span>
<span><span class="va">membership_dir</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">'extdata'</span>, <span class="st">'seacells_cell_membership_rna_syn.csv'</span>, package <span class="op">=</span> <span class="st">'mcRigor'</span><span class="op">)</span></span>
<span><span class="va">cell_membership</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.csv</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">membership_dir</span>, check.names <span class="op">=</span> <span class="cn">F</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-1-identify-single-cells-that-constitute-the-dubious-metacells">Step 1: identify single cells that constitute the dubious
metacells<a class="anchor" aria-label="anchor" href="#step-1-identify-single-cells-that-constitute-the-dubious-metacells"></a>
</h2>
<p>In this tutorial, we focus on the metacell partition corresponding to
the granularity level
<math display="inline" xmlns="http://www.w3.org/1998/Math/MathML"><semantics><mrow><mi>γ</mi><mo>=</mo><mn>50</mn></mrow><annotation encoding="application/x-tex">\gamma = 50</annotation></semantics></math>,
specified by <code>tgamma</code> = 50. We first run the main mcRigor
function <code>mcRigor_DETECT</code> to check whether this partition
contains any dubious metacells. If no dubious metacells are detected,
there is no need to apply mcRigor two-step.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tgamma</span> <span class="op">=</span> <span class="fl">50</span></span>
<span><span class="va">detect_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/mcRigor_DETECT.html">mcRigor_DETECT</a></span><span class="op">(</span>obj_singlecell <span class="op">=</span> <span class="va">obj_singlecell</span>, cell_membership <span class="op">=</span> <span class="va">cell_membership</span>, tgamma <span class="op">=</span> <span class="va">tgamma</span><span class="op">)</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/table.html" class="external-link">table</a></span><span class="op">(</span><span class="va">detect_res</span><span class="op">$</span><span class="va">mc_res</span><span class="op">)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;     dubious trustworthy </span></span>
<span><span class="co">#&gt;          28         240</span></span></code></pre></div>
<p>We proceed with mcRigor two-step because dubious metacells exist in
this partition We call the function <code>mcRigorTS_Step1</code> to
identify the single cells that constitute dubious metacells (under a
lower divergence score threshold) and need to be re-partitioned. Note
that we input the permutation results, <code>detect_res$TabMC</code>,
obtained from the previous code block to save computation time here,
although it is also acceptable to leave this argument as its default
value (<code>NULL</code>). We need to specify the metacell partitioning
method (e.g., SEACells, MetaCell, MetaCell2, SuperCell, or MetaQ) by
setting the argument <code>method</code>, ensuring that the function
generates outputs in the correct format.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">sc_membership</span> <span class="op">=</span> <span class="va">cell_membership</span><span class="op">[[</span><span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">tgamma</span><span class="op">)</span><span class="op">]</span><span class="op">]</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">sc_membership</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">cell_membership</span><span class="op">)</span></span>
<span></span>
<span><span class="va">step1_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/mcRigorTS_Step1.html">mcRigorTS_Step1</a></span><span class="op">(</span>obj_singlecell <span class="op">=</span> <span class="va">obj_singlecell</span>, sc_membership <span class="op">=</span> <span class="va">sc_membership</span>, TabMC <span class="op">=</span> <span class="va">detect_res</span><span class="op">$</span><span class="va">TabMC</span>, method <span class="op">=</span> <span class="st">'seacells'</span><span class="op">)</span></span>
<span><span class="co">#&gt; Testing thresholds (Thre) not provided. Calculating thresholds with defaults...</span></span></code></pre></div>
<p>The Seurat object of single cells that need to be re-partitioned is
stored in the <code>obj_sc_dub</code> field of the output
<code>step1_res</code>. If the metacell partitioning method used is
SEACells, MetaCell2, or MetaQ,, <code>mcRigorTS_Step1</code> will
automatically generate two CSV files (<code>counts.csv</code> and
<code>metadata.csv</code>) in the current working directory. These files
can be directly used as input for the corresponding metacell
partitioning methods in Python.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">step1_res</span><span class="op">$</span><span class="va">obj_sc_dub</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 1982 features across 2883 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (1982 features, 1982 variable features)</span></span>
<span><span class="co">#&gt;  3 layers present: counts, data, scale.data</span></span>
<span><span class="co">#&gt;  2 dimensional reductions calculated: pca, umap</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="st">"seacells/"</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "counts.csv"   "metadata.csv"</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="step-2-re-organize-the-single-cells-into-more-trustworthy-metacells">Step 2: Re-organize the single cells into more trustworthy
metacells<a class="anchor" aria-label="anchor" href="#step-2-re-organize-the-single-cells-into-more-trustworthy-metacells"></a>
</h2>
<p>We then need to apply the chosen metacell partitioning method (e.g.,
SEACells, MetaCell, MetaCell2, SuperCell, MetaQ) to these identified
single cells to obtain more trustworthy metacell partitions for them,
under granularity level smaller than <code>tgamma</code> (recommend
obtaining all the candidate paritions corresponding to granularity
levels from 2 to <code>tgamma</code>-1). One can refer to the <a href="https://jsb-ucla.github.io/mcRigor/articles/mcRigor-3-mc-method.html">Implementing
metacell partitioning methods</a> tutorial for more guidnance on how to
obtain the metacell partitions. The candidate metacell partitions can be
stored in csv file and read into an R object
<code>cell_membership_twostep</code> using
<code>cell_membership_twostep &lt;- read.csv(file = "/path/to/the/csv/file", check.names = F, row.names = 1)</code>.
For illustration purpose in this tutorial, we set
<code>cell_membership_twostep</code> as the partitions corresponding to
small granularity levels in <code>cell_membership_all</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">cell_membership_twostep</span> <span class="op">=</span> <span class="va">cell_membership</span><span class="op">[</span> , <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">cell_membership</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;</span> <span class="fl">50</span><span class="op">]</span></span>
<span><span class="va">cell_membership_twostep</span> <span class="op">=</span> <span class="va">cell_membership_twostep</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">step1_res</span><span class="op">$</span><span class="va">obj_sc_dub</span><span class="op">)</span>,<span class="op">]</span></span></code></pre></div>
<p>We call the function <code>mcRigorTS_Step2</code> to select the
optimal granularity level and output the final metacell partition with
remaining dubious metacells flagged.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">step2_res</span> <span class="op">=</span> <span class="fu"><a href="../reference/mcRigorTS_Step2.html">mcRigorTS_Step2</a></span><span class="op">(</span>step1_res <span class="op">=</span> <span class="va">step1_res</span>, obj_singlecell <span class="op">=</span> <span class="va">obj_singlecell</span>, cell_membership_twostep <span class="op">=</span> <span class="va">cell_membership_twostep</span><span class="op">)</span></span>
<span><span class="co">#&gt; Names of identity class contain underscores ('_'), replacing with dashes ('-')</span></span>
<span><span class="co">#&gt; First group.by variable `Metacell` starts with a number, appending `g` to ensure valid variable names</span></span>
<span><span class="co">#&gt; <span style="color: #555555;">This message is displayed once every 8 hours.</span></span></span></code></pre></div>
<p>The Seurat object of the final metacells are stored in the
<code>obj_metacell_final</code> field of the output
<code>step2_res</code> with the dubious metacell detection results
recorded in its metadata under the name <code>mcRigor</code>. Another
field, <code>obj_metacell_final_withsc</code>, is the Seurat object of
the final metacells where all dubious metacells have been dissected into
their constituent single cells (i.e., no dubious metacells remain).
Users can choose which object to use depending on their analytical
needs. In addtion, the output <code>step2_res</code> includes two
visualization fields: <code>plot</code> and <code>plot_withsc</code>,
which correspond to UMAP plots of single cells with
<code>obj_metacell_final</code> and
<code>obj_metacell_final_withsc</code> metacells projected onto the UMAP
space, respectively.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">step2_res</span><span class="op">$</span><span class="va">obj_metacell_final</span></span>
<span><span class="co">#&gt; An object of class Seurat </span></span>
<span><span class="co">#&gt; 2000 features across 311 samples within 1 assay </span></span>
<span><span class="co">#&gt; Active assay: RNA (2000 features, 0 variable features)</span></span>
<span><span class="co">#&gt;  2 layers present: counts, data</span></span>
<span><span class="va">step2_res</span><span class="op">$</span><span class="va">plot</span></span></code></pre></div>
<p><img src="mcRigor-4-twostep_files/figure-html/unnamed-chunk-9-1.png" width="700"></p>
</div>
<div class="section level2">
<h2 id="session-information">Session information<a class="anchor" aria-label="anchor" href="#session-information"></a>
</h2>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; R version 4.5.1 (2025-06-13)</span></span>
<span><span class="co">#&gt; Platform: x86_64-pc-linux-gnu</span></span>
<span><span class="co">#&gt; Running under: Ubuntu 24.04.3 LTS</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Matrix products: default</span></span>
<span><span class="co">#&gt; BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3 </span></span>
<span><span class="co">#&gt; LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/libopenblasp-r0.3.26.so;  LAPACK version 3.12.0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; locale:</span></span>
<span><span class="co">#&gt;  [1] LC_CTYPE=C.UTF-8       LC_NUMERIC=C           LC_TIME=C.UTF-8       </span></span>
<span><span class="co">#&gt;  [4] LC_COLLATE=C.UTF-8     LC_MONETARY=C.UTF-8    LC_MESSAGES=C.UTF-8   </span></span>
<span><span class="co">#&gt;  [7] LC_PAPER=C.UTF-8       LC_NAME=C              LC_ADDRESS=C          </span></span>
<span><span class="co">#&gt; [10] LC_TELEPHONE=C         LC_MEASUREMENT=C.UTF-8 LC_IDENTIFICATION=C   </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; time zone: UTC</span></span>
<span><span class="co">#&gt; tzcode source: system (glibc)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; attached base packages:</span></span>
<span><span class="co">#&gt; [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; other attached packages:</span></span>
<span><span class="co">#&gt; [1] ggplot2_4.0.0      Seurat_5.3.0       SeuratObject_5.2.0 sp_2.2-0          </span></span>
<span><span class="co">#&gt; [5] mcRigor_1.0        BiocStyle_2.36.0  </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; loaded via a namespace (and not attached):</span></span>
<span><span class="co">#&gt;   [1] deldir_2.0-4           pbapply_1.7-4          gridExtra_2.3         </span></span>
<span><span class="co">#&gt;   [4] rlang_1.1.6            magrittr_2.0.4         RcppAnnoy_0.0.22      </span></span>
<span><span class="co">#&gt;   [7] spatstat.geom_3.6-0    matrixStats_1.5.0      ggridges_0.5.7        </span></span>
<span><span class="co">#&gt;  [10] compiler_4.5.1         reshape2_1.4.4         png_0.1-8             </span></span>
<span><span class="co">#&gt;  [13] systemfonts_1.3.1      vctrs_0.6.5            stringr_1.5.2         </span></span>
<span><span class="co">#&gt;  [16] pkgconfig_2.0.3        fastmap_1.2.0          labeling_0.4.3        </span></span>
<span><span class="co">#&gt;  [19] promises_1.3.3         rmarkdown_2.30         ragg_1.5.0            </span></span>
<span><span class="co">#&gt;  [22] purrr_1.1.0            xfun_0.53              cachem_1.1.0          </span></span>
<span><span class="co">#&gt;  [25] jsonlite_2.0.0         goftest_1.2-3          later_1.4.4           </span></span>
<span><span class="co">#&gt;  [28] spatstat.utils_3.2-0   irlba_2.3.5.1          parallel_4.5.1        </span></span>
<span><span class="co">#&gt;  [31] cluster_2.1.8.1        R6_2.6.1               ica_1.0-3             </span></span>
<span><span class="co">#&gt;  [34] spatstat.data_3.1-8    stringi_1.8.7          bslib_0.9.0           </span></span>
<span><span class="co">#&gt;  [37] RColorBrewer_1.1-3     reticulate_1.43.0      spatstat.univar_3.1-4 </span></span>
<span><span class="co">#&gt;  [40] parallelly_1.45.1      lmtest_0.9-40          jquerylib_0.1.4       </span></span>
<span><span class="co">#&gt;  [43] scattermore_1.2        Rcpp_1.1.0             bookdown_0.45         </span></span>
<span><span class="co">#&gt;  [46] knitr_1.50             tensor_1.5.1           future.apply_1.20.0   </span></span>
<span><span class="co">#&gt;  [49] zoo_1.8-14             sctransform_0.4.2      httpuv_1.6.16         </span></span>
<span><span class="co">#&gt;  [52] Matrix_1.7-3           splines_4.5.1          igraph_2.2.0          </span></span>
<span><span class="co">#&gt;  [55] tidyselect_1.2.1       abind_1.4-8            yaml_2.3.10           </span></span>
<span><span class="co">#&gt;  [58] spatstat.random_3.4-2  spatstat.explore_3.5-3 codetools_0.2-20      </span></span>
<span><span class="co">#&gt;  [61] miniUI_0.1.2           listenv_0.9.1          plyr_1.8.9            </span></span>
<span><span class="co">#&gt;  [64] lattice_0.22-7         tibble_3.3.0           withr_3.0.2           </span></span>
<span><span class="co">#&gt;  [67] shiny_1.11.1           S7_0.2.0               ROCR_1.0-11           </span></span>
<span><span class="co">#&gt;  [70] evaluate_1.0.5         Rtsne_0.17             future_1.67.0         </span></span>
<span><span class="co">#&gt;  [73] fastDummies_1.7.5      desc_1.4.3             survival_3.8-3        </span></span>
<span><span class="co">#&gt;  [76] polyclip_1.10-7        fitdistrplus_1.2-4     pillar_1.11.1         </span></span>
<span><span class="co">#&gt;  [79] BiocManager_1.30.26    KernSmooth_2.23-26     plotly_4.11.0         </span></span>
<span><span class="co">#&gt;  [82] generics_0.1.4         RcppHNSW_0.6.0         scales_1.4.0          </span></span>
<span><span class="co">#&gt;  [85] globals_0.18.0         xtable_1.8-4           glue_1.8.0            </span></span>
<span><span class="co">#&gt;  [88] lazyeval_0.2.2         tools_4.5.1            data.table_1.17.8     </span></span>
<span><span class="co">#&gt;  [91] RSpectra_0.16-2        RANN_2.6.2             fs_1.6.6              </span></span>
<span><span class="co">#&gt;  [94] dotCall64_1.2          cowplot_1.2.0          grid_4.5.1            </span></span>
<span><span class="co">#&gt;  [97] tidyr_1.3.1            nlme_3.1-168           patchwork_1.3.2       </span></span>
<span><span class="co">#&gt; [100] cli_3.6.5              spatstat.sparse_3.1-0  textshaping_1.0.4     </span></span>
<span><span class="co">#&gt; [103] spam_2.11-1            viridisLite_0.4.2      dplyr_1.1.4           </span></span>
<span><span class="co">#&gt; [106] uwot_0.2.3             gtable_0.3.6           sass_0.4.10           </span></span>
<span><span class="co">#&gt; [109] digest_0.6.37          progressr_0.17.0       ggrepel_0.9.6         </span></span>
<span><span class="co">#&gt; [112] htmlwidgets_1.6.4      farver_2.1.2           htmltools_0.5.8.1     </span></span>
<span><span class="co">#&gt; [115] pkgdown_2.1.3          lifecycle_1.0.4        httr_1.4.7            </span></span>
<span><span class="co">#&gt; [118] mime_0.13              MASS_7.3-65</span></span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc" aria-label="Table of contents"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Pan Liu.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
